packages/block-library/src/link-preview/edit.js

```diff
@@ -0,0 +1,131 @@
+/**
+ * WordPress dependencies
+ */
+import { __, _x } from '@wordpress/i18n';
+import { useState, useEffect } from '@wordpress/element';
+import { useBlockProps, BlockControls } from '@wordpress/block-editor';
+import { __experimentalFetchUrlData } from '@wordpress/core-data';
+import {
+	Placeholder,
+	Spinner,
+	Button,
+	ToolbarGroup,
+	ToolbarButton,
+	Notice,
+} from '@wordpress/components';
+import { link, edit } from '@wordpress/icons';
+
+/**
+ * Internal dependencies
+ */
+import { Content } from './content';
+
+export default function LinkPreviewEdit( props ) {
+	const { attributes, setAttributes } = props;
+	const { url, title } = attributes;
+	const [ isFetching, setIsFetching ] = useState( false );
+	const [ isEditingUrl, setIsEditingUrl ] = useState( ! url );
+	const [ urlValue, setURLValue ] = useState( url );
+	const [ hasError, setHasError ] = useState( false );
+
+	const blockProps = useBlockProps( {
+		onClick: isEditingUrl
+			? undefined
+			: ( event ) => {
+					event.preventDefault();
+			  },
+	} );
+
+	useEffect( () => {
+		if ( url && ! title ) {
+			setIsFetching( true );
+			__experimentalFetchUrlData( url )
+				.catch( () => {
+					setHasError( true );
+				} )
+				.then( ( data ) => {
+					if ( ! data || ! data.title ) {
+						setHasError( true );
+					} else {
+						setHasError( false );
+						setIsEditingUrl( false );
+						setAttributes( {
```

Comment:
I think we also need a couple of things if this effect is required.

- Add `__unstableMarkNextChangeAsNotPersistent` before `setAttributes` to prevent "[undo traps](https://github.com/WordPress/gutenberg/issues/8119#issuecomment-1185238067)".
- Add a cleanup function to prevent race conditions and memory leaks. Don't update states if the component is already unmounted.